model/task_queue_service_dependency.go

```diff
@@ -457,15 +436,59 @@ func (d *basicCachedDAGDispatcherImpl) FindNextTask(ctx context.Context, spec Ta
 	return nil
 }
 
+// tryMarkItemDispatched will dispatch a standalone task if all of the following are true:
+// (a) it's not marked as dispatched in the in-memory queue.
+// (b) a record of the task exists in the database.
+// (c) it never previously ran on another host.
+// (d) all of its dependencies are satisfied.
+// Returns true if the item has already been marked dispatched by a separate request.
+func (d *basicCachedDAGDispatcherImpl) tryMarkItemDispatched(item *TaskQueueItem) bool {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	if item.IsDispatched {
+		return true
+	}
+
+	// Cache the task as dispatched from the in-memory queue's point of view.
+	// However, it won't actually be dispatched to a host if it doesn't satisfy all constraints.
+	item.IsDispatched = true // *TaskQueueItem
+	return false
+}
+
+func (d *basicCachedDAGDispatcherImpl) getNextTaskForTaskGroup(taskGroupUnit schedulableUnit) *TaskQueueItem {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	next := d.nextTaskGroupTask(taskGroupUnit)
+	if next != nil {
+		// next is a *TaskQueueItem, sourced for d.taskGroups (map[string]schedulableUnit) tasks' field, which in turn is a []TaskQueueItem.
+		// taskGroupTask is a *TaskQueueItem sourced from d.nodeItemMap, which is a map[node.ID()]*TaskQueueItem.
+		node := d.getNodeByItemID(next.Id)
+		taskGroupTask := d.getItemByNodeID(node.ID())
+		taskGroupTask.IsDispatched = true
```

Comment:
Does this have to check `item.IsDispatched` before dispatching it? It seems like it should be doing a similar (or possibly identical) check to `tryMarkItemDispatched` to avoid two hosts running the same task group from dispatching the same task.

Reply:
Yes, it does [here](https://github.com/evergreen-ci/evergreen/blob/003a08f1f323d0ea0c90e944a2fbb5c5c799ee70/model/task_queue_service_dependency.go#L566) in `nextTaskGroupTask`